unpacker.go

```diff
@@ -87,85 +88,127 @@ func (u *unpacker) unpack(ctx context.Context, config ocispec.Descriptor, layers
 		a  = u.c.DiffService()
 		cs = u.c.ContentStore()
 
-		states []layerState
-		chain  []digest.Digest
+		chain []digest.Digest
+
+		fetchOffset int
+		fetchC      []chan struct{}
+		fetchErr    chan error
 	)
+
+	// If there is an early return, ensure any ongoing
+	// fetches get their context cancelled
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+EachLayer:
 	for i, desc := range layers {
-		states = append(states, layerState{
-			layer: rootfs.Layer{
-				Blob: desc,
-				Diff: ocispec.Descriptor{
-					MediaType: ocispec.MediaTypeImageLayer,
-					Digest:    diffIDs[i],
-				},
-			},
-		})
-	}
-	for {
-		var layer ocispec.Descriptor
-		select {
-		case layer = <-u.updateCh:
-		case <-ctx.Done():
-			return ctx.Err()
+		parent := identity.ChainID(chain)
+		chain = append(chain, diffIDs[i])
+
+		chainID := identity.ChainID(chain).String()
+		if _, err := sn.Stat(ctx, chainID); err == nil {
```

Comment:
Can we stat from top layer to bot layer, and break once a snapshot is found?

That is an optimization, and make it possible to support squashed remote snapshots. :)

Reply:
That did cross my mind, you make things so much harder ;)

I'll look at that tomorrow

Reply:
@Random-Liu would merging it like this today break any snapshot implementations? As we discussed yesterday, figuring out how to handle the parentage for squashing is kind of tricky. For example we would need to ensure that while referencing a chain id, we aren't enforcing the parent. We currently don't enforce this on the client but the metadata store is.

Reply:
@Random-Liu @dmcgowan Even without that optimization, we can support squashed remote snapshots using a digest of a manifest (discussed with CernVM-FS people in https://github.com/ktock/remote-snapshotter/issues/7 ). We will be able to pass these digests to remote snapshotters by PR https://github.com/containerd/containerd/pull/3846#issuecomment-564061264 .